### PR TITLE
Feature/rbsfxirjtp 168 orderbook service aggregation of incoming updates into a bid/ask orderbook

### PR DESCRIPTION
**Pull request by pashunov151**

### CODE REVIEW

**Review by valto2 — DISMISSED** [review]

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 1]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
         allItems.remove(itemName);
     }
 
+    public void updateItem(Aurora.AuroraResponse response) {
+        if (response.getMessage().is(TickResponse.class)) {
```
Comment (gaantonova) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:58]: please refactor so, that we throw an exception here, and remove the else clause

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 2]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+            item.setPrice(tickResponse.getPrice());
+            item.setQuantity(tickResponse.getQuantity());
+            item.setOrigin(tickResponse.getOrigin());
+
+            productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+            Set<Item> itemSet = allItems.get(tickResponse.getGoodName());
+            if (itemSet != null) {
+                if (itemSet.contains(item)) {
+                    Item presentItem = itemSet.stream().filter(item1 -> Double.compare(item1.getPrice(), item.getPrice()) == 0
```
Comment (gaantonova) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:77]: item1 is not a good nme, please change it

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 3]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+        if (response.getMessage().is(TickResponse.class)) {
+            TickResponse tickResponse;
+
+            try {
+                tickResponse = response.getMessage().unpack(TickResponse.class);
+            } catch (InvalidProtocolBufferException e) {
+                throw new IncorrectResponseException("Response is not correct!");
+            }
+
+            Item item = new Item();
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:67]: this item population could be extracted in a separate metode

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            }
+            LOGGER.info("Forwarding to orderBookService.cancelItemSubscription method");
+
+            orderBookService.cancelItemSubscription(request, responseObserver);
+        }
+    }
+
+    @Override
+    public void subscribe(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        responseObserver.onError(Status.UNIMPLEMENTED.withDescription("Method is not implemented").asException());
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:68]: I think it would be better to have a class with constants (or messages) that are common for the whole service and "Method is not implemented" to be extracted in that separate class, following some pattern for messages thrown from the service

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 5]
```diff
@@ -37,24 +34,28 @@ private OrderBookService(final AuroraClient auroraClient, final ItemMarket itemM
...
-                    .build());
+    public void getOrderBookItemLayers(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        String[] topicSplit = request.getTopic().split(REGEX);
+        String itemName = topicSplit[1];
+        long itemQuantity = Long.parseLong(topicSplit[2]);
+
+        List<OrderBookLayer> requestedItem = itemMarket.getRequestedItem(itemName, itemQuantity);
+
+        if (requestedItem.size() > 0) {
+            responseObserver.onNext(
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:45]: both if and else are doing absolutely the same thing, it could be extracted as a method and to pass ItemOrderBookResponse as a parameter

File: order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java [hunk 6]
```diff
@@ -32,6 +40,73 @@ public void setUp() {
...
+    @Test
+    public void updateItemThrowsExceptionWhenPassingRequestOfDifferentType() {
+        //Arrange
+        itemMarket.addTrackedItem("cheese");
+        TickResponse cheese = TickResponse.newBuilder().setGoodName("cheese").setQuantity(2).setPrice(2.6).setOrigin(Origin.ASIA).build();
+        Aurora.AuroraResponse build = Aurora.AuroraResponse.newBuilder().setMessage(Any.pack(cheese)).build();
+
+
+        //Act
+        itemMarket.updateItem(build);
```
Comment (zaraandreevaa) [inline order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java:79]: why aren't we asserting or expecting the exception?

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 7]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+            item.setPrice(tickResponse.getPrice());
+            item.setQuantity(tickResponse.getQuantity());
+            item.setOrigin(tickResponse.getOrigin());
+
+            productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:72]: it would be better to use this.productsQuantity

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 8]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+
+            Set<Item> itemSet = allItems.get(tickResponse.getGoodName());
+            if (itemSet != null) {
+                if (itemSet.contains(item)) {
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:76]: could we merge the 2 if statements as itemSet != null && itemSet.contains(item)

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 9]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+                    currentItem.setQuantity(currentItem.getQuantity() - itemLeft);
+                } else if (currentItem.getQuantity() < itemLeft) {
+                    currentLayer.setQuantity(currentItem.getQuantity());
+
+                    iterator.remove();
+                }
+                productsQuantity.put(itemName, productsQuantity.get(itemName) - currentLayer.getQuantity());
+                itemLeft -= currentLayer.getQuantity();
+
+                OrderBookLayer build = currentLayer
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:134]: orderBookLayer would be a better name for the object

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 10]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+
+            layers = new ArrayList<>();
+
+            Iterator<Item> iterator = items.iterator();
+            long itemLeft = quantity;
+
+            while (iterator.hasNext() && itemLeft > 0) {
+                Item currentItem = iterator.next();
+
+                OrderBookLayer.Builder currentLayer = OrderBookLayer.newBuilder()
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:115]: all this logic from like 115 to line 130 is not directly related to getRequestedItem, it could be in a separate method

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+            if (command.equals(SUBSCRIBE)) {
+                LOGGER.info("Forwarding to orderBookService.announceItemInterest method");
+
+                orderBookService.announceItemInterest(request, responseObserver);
+                return;
+            }
+            LOGGER.info("Forwarding to orderBookService.cancelItemSubscription method");
+
+            orderBookService.cancelItemSubscription(request, responseObserver);
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:62]: more like a business question - are we still suppose to call cancelItemSubscription after executing announceItemInterest?

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    private static final String NUMBER_CHECK_REGEX = "^\\d+$";
+
+    private static final String SUBSCRIBE = "subscribe";
+
+    private static final Pattern NUMBER_CHECK_PATTERN =
+            Pattern.compile(NUMBER_CHECK_REGEX);
+
+
+    @Autowired
+    public AuroraRequestHandler(OrderBookService orderBookService) {
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:34]: is this constructor consistent with other classes' implementation? I believe there was usage of lombok somewhere - this should be approved by the team - are you using constructors or lombok

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 1 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:58]: Done

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 2 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:77]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 3 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:67]: Method extracted.

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java (hunk 4 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:68]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java (hunk 5 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:45]: Fixed

File: order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java (hunk 6 above)
Comment (valto2) [reply order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java:79]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 7 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:72]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 8 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:76]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 9 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:134]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 10 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:115]: Done

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java (hunk 11 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:62]: Yes, we are suppose to call the method , because it is one of the cases.

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java (hunk 12 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:34]: Fixed

Comment (rusevradoslav) [reply order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java:34]: You can use @RequiredArgsConstructor(onConstructor = @__(@Autowired)), but the whole team follow the convention with constructors. It should be discussed with other team members.

File: order-book/src/main/java/com/market/banica/order/book/helper/ConstantsAuroraRequestHandler.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package com.market.banica.order.book.helper;
+
+import java.util.regex.Pattern;
+
+public class ConstantsAuroraRequestHandler {
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/helper/ConstantsAuroraRequestHandler.java:5]: the pattern for naming class with constants is Constants to be at the end - like for the tests. So better naming here would be the module name + Constants (this is because the class could be used across the whole module and should be common for all classes)

Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/helper/ConstantsAuroraRequestHandler.java]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 14]
```diff
@@ -1,58 +1,179 @@
...
+
+        if (currentItem.getQuantity() >= itemLeft) {
+            currentLayer.setQuantity(itemLeft);
+
+            if (currentItem.getQuantity() == itemLeft) {
+                iterator.remove();
+            }
+
+            currentItem.setQuantity(currentItem.getQuantity() - itemLeft);
+        } else if (currentItem.getQuantity() < itemLeft) {
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:139]: the if is not necessary here, the first if is checking if it is >= then the else will be executed for <

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 15]
```diff
@@ -64,35 +47,37 @@ public void getOrderBookItemLayers(ItemOrderBookRequest request, StreamObserver<
...
     }
 
     @Override
     public void cancelItemSubscription(CancelSubscriptionRequest request, StreamObserver<CancelSubscriptionResponse> responseObserver) {
+        String itemName = request.getItemName();
 
         try {
 
-            auroraClient.stopSubscription(request.getItemName(), request.getClientId());
+            auroraClient.stopSubscription(itemName, request.getClientId());
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:73]: if you extracted the item name because it is repeated twice, you should do it also for the clientId

File: order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java [hunk 16]
```diff
@@ -86,36 +90,36 @@ public void setUp() {
...
+
         //Assert
         assertEquals(0, bookItemLayers.getOrderbookLayersList().size());
     }
 
     @Test
     public void announceItemInterestExecutesSuccessfullyWithValidInterestRequest() {
-        //Act
-        InterestsResponse interestsResponse = blockingStub.announceItemInterest(INTERESTS_REQUEST);
+        //Act interestsResponse
```
Comment (zaraandreevaa) [inline order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java:121]: interestsResponse is added by mistake?

File: order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java [hunk 17]
```diff
@@ -86,36 +90,36 @@ public void setUp() {
...
     @Test
-    public void getOrderBookItemLayersReturnsAnEmptyItemOrderBookResponseWhenItemMarketDoesNotContainRequestItemName() {
-        //Arrange
-        when(itemMarket.getItemSetByName(any())).thenReturn(Optional.empty());
-
+    public void getOrderBookItemLayersReturnsAnEmptyItemOrderBookResponseWhenItemMarketDoesNotContainRequestItemName() throws InvalidProtocolBufferException {
         //Act
         ItemOrderBookResponse bookItemLayers = blockingStub.getOrderBookItemLayers(ITEM_ORDER_BOOK_REQUEST);
 
+
```
Comment (zaraandreevaa) [inline order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java:114]: 2 empty lines between Act and Assert?

File: order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java [hunk 18]
```diff
@@ -86,36 +90,36 @@ public void setUp() {
...
-    public void getOrderBookItemLayersSuccessfullyReturnsItemOrderBookResponseWhenItemMarketContainsRequestItemName() {
+    public void getOrderBookItemLayersSuccessfullyReturnsItemOrderBookResponseWhenItemMarketContainsRequestItemName() throws InvalidProtocolBufferException {
         //Arrange
-        when(itemMarket.getItemSetByName(any())).thenReturn(Optional.of(this.items));
+        when(itemMarket.getRequestedItem(any(), any(Long.class))).thenReturn(orderBookLayers);
 
         //Act
         ItemOrderBookResponse bookItemLayers = blockingStub.getOrderBookItemLayers(ITEM_ORDER_BOOK_REQUEST);
 
+
```
Comment (zaraandreevaa) [inline order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java:100]: 2 empty lines between Act and Assert?

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 14 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:139]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java (hunk 15 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:73]: Fixed

File: order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java (hunk 16 above)
Comment (valto2) [reply order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java:121]: Fixed

File: order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java (hunk 17 above)
Comment (valto2) [reply order-book/src/test/java/com/market/banica/order/book/service/grpc/OrderBookServiceTest.java:114]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 19]
```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
...
+        List<OrderBookLayer> layers;
+        try {
+            lock.writeLock().lock();
+
+            layers = new ArrayList<>();
+
+            Iterator<Item> iterator = items.iterator();
+            long itemLeft = quantity;
+
+            while (iterator.hasNext() && itemLeft > 0) {
```
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:108]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java]: Already discussed and resolved

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 20]
```diff
@@ -37,24 +34,28 @@ private OrderBookService(final AuroraClient auroraClient, final ItemMarket itemM
...
-                            .map(item -> OrderBookLayer.newBuilder()
-                                    .setPrice(item.getPrice())
-                                    .setQuantity(item.getQuantity())
-                                    .setOrigin(item.getOrigin())
-                                    .build()).collect(Collectors.toList()))
-                    .build());
+    public void getOrderBookItemLayers(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        String[] topicSplit = request.getTopic().split(REGEX);
+        String itemName = topicSplit[1];
+        long itemQuantity = Long.parseLong(topicSplit[2]);
```
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 21]
```diff
@@ -63,36 +64,43 @@ public void getOrderBookItemLayers(ItemOrderBookRequest request, StreamObserver<
...
             responseObserver.onError(Status.INVALID_ARGUMENT.withDescription("Invalid item name").asException());
 
         }
     }
 
     @Override
-    public void cancelItemSubscription(CancelSubscriptionRequest request, StreamObserver<CancelSubscriptionResponse> responseObserver) {
+    public void cancelItemSubscription(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        String[] topicSplit = request.getTopic().split(REGEX);
+        String itemName = topicSplit[1];
```
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 22]
```diff
@@ -63,36 +64,43 @@ public void getOrderBookItemLayers(ItemOrderBookRequest request, StreamObserver<
     }
 
     @Override
-    public void announceItemInterest(InterestsRequest request, StreamObserver<InterestsResponse> responseObserver) {
-
+    public void announceItemInterest(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        String[] topicSplit = request.getTopic().split(REGEX);
+        String itemName = topicSplit[1];
```
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public void request(Aurora.AuroraRequest request, StreamObserver<Aurora.AuroraResponse> responseObserver) {
+        String[] topicSplit = request.getTopic().split(SPLIT_SLASH_REGEX);
+        String prefix = topicSplit[0];
+        if (!prefix.equals("orderbook")) {
+            LOGGER.debug("Not valid request, starts with: {}, expected: 'orderbook'", prefix);
+            responseObserver.onError(Status.INVALID_ARGUMENT.withDescription("Invalid request").asException());
+            return;
+        }
+        if (topicSplit.length == 3 && isValidNumber(topicSplit[2])) {
```
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/AuroraRequestHandler.java]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java]: We are already tracking the item, then we recieve response from the market and then we are adding the quantity if that's what you are asking.

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java [hunk 24]
```diff
@@ -64,35 +47,39 @@ public void getOrderBookItemLayers(ItemOrderBookRequest request, StreamObserver<
 
     @Override
     public void announceItemInterest(InterestsRequest request, StreamObserver<InterestsResponse> responseObserver) {
+        final String itemName = request.getItemName();
 
         try {
 
-            auroraClient.startSubscription(request.getItemName(), request.getClientId());
+            auroraClient.startSubscription(itemName, request.getClientId());
```
Comment (zaraandreevaa) [inline order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:54]: probably the request.clientId should be also extracted as the itemName

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gaantonova) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java]: yes, you've changed it, it  used to be before the clause where you check if you track it

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java [hunk 25]
```diff
@@ -1,58 +1,179 @@
...
+        Item item = populateItem(tickResponse);
+
+        this.productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+        LOGGER.info("Products data updated!");
+
+        if (itemSet.contains(item)) {
+
+            Item presentItem = itemSet.stream().filter(currentItem -> Double.compare(currentItem.getPrice(), item.getPrice()) == 0
+                    && currentItem.getOrigin().equals(item.getOrigin())).findFirst().get();
```
Comment (gaantonova) [inline order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:85]: please use item.compareTo, as the functionality is the same

File: order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java [hunk 26]
```diff
@@ -14,24 +21,94 @@
...
+        TickResponse cheese3 = TickResponse.newBuilder().setGoodName(CHEESE_ITEM_NAME).setQuantity(2).setPrice(2.6).setOrigin(Origin.AMERICA).build();
+        Aurora.AuroraResponse build3 = Aurora.AuroraResponse.newBuilder().setMessage(Any.pack(cheese3)).build();
+
+        //Act
+        itemMarket.updateItem(build);
+        itemMarket.updateItem(build2);
+        itemMarket.updateItem(build3);
+
+        //Assert
+        assertEquals(2, allItems.get(CHEESE_ITEM_NAME).size());
```
Comment (gaantonova) [inline order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java:67]: a good check will be also to determine what is the quantity in the item itself, as we duplicate the information in products quantity, we need to be sure that the values match

File: order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java (hunk 24 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/service/grpc/OrderBookService.java:54]: Fixed

File: order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java (hunk 26 above)
Comment (valto2) [reply order-book/src/test/java/com/market/banica/order/book/model/ItemMarketTest.java:67]: Fixed

File: order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java (hunk 25 above)
Comment (valto2) [reply order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java:85]: Fixed

**Review by zaraandreevaa — APPROVED** [review]

**Review by gaantonova — APPROVED** [review]
